Add section to documentation with howto for AcroForm fields extraction

typhoon71 opened the pull request:
**Pull request**

- This PR adds an AcroForm fields extraction howto, previosly missing in the docs.
- It adds a working code snippet that extracts all fields from AcroForm PDF.
- References issue #182.
- Fixes #336 
- Fixes #293

**How Has This Been Tested?**

It was tested on a sample of complex PDF files and it correctly extracts all fields.
It is not possible to share those PDF files, any idea how to provide samples?

**Checklist**

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added docstrings to newly created methods and classes
- [x] I have optimized the code at least one time after creating the initial version
- [x] I have updated the [README.md](../README.md) or I have verified that this is not necessary
- [x] I have updated the [readthedocs](../docs/source) documentation or I verified that this is not necessary
- [x] I have added a consice human-readable description of the change to [CHANGELOG.md](../CHANGELOG.md)

pietermarsman commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+
+Before you start, make sure you have :ref:`installed pdfminer.six<install>`.
+
+The second thing you need is a PDF with AcroForms (as found in PDF files with fillable forms or multiple choices).
```
While howto's are aimed at people that want to get things done (and thus probably have a PDF with acroforms at their hands), there might also be an audience that is following this tutorial to learn how to work with pdfminer.six. These people can benefit from an example pdf with an AcroForm. Maybe add one to the /samples directory and add a downloadable link here. 

Also, how do people know if there PDF has an AcroForm? Maybe you can include a section with method that can be used to check if the PDF has one.

typhoon71 replied:
My problem is I can't share my PDF files... I'll see how to create some sinple ones or find examples online.

A simple method to check if there are forms would just be a check:
```python
if 'AcroForm' in doc.catalog:
```
So basically it's already there. I could change the code to return None if there's no AcroForm and make this code a method.

jstockwin replied:
If it's that simple, I would just add a sentence or two at the top. "To check if you're PDF has an AcroForm, you can simply run....".

pietermarsman commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+                    if isinstance(value, bytes):
+                        value = utils.decode_text(value)
+              
+                data.update({name: value})    
```
I think it makes sense to extract a method that has the `PDFDocument` as an input and returns the `data` object. Then we give the reader a piece of code that they can easily reuse / map over multiple documents.

typhoon71 replied:
Oh, makes sense, I'll see about it.

Question!

You may remember I was lamenting pdfminer was slow when creating the `PDFDocument`.
I found the reason: basically when creating `PDFDocument`all of the ObjRef are being resolved.

For extracting the AcroForm fields this is overkill, at least for my forms (where there are lots of unrelated ObjRef).
It should be better to get the catalog, check if there's an `AcroForm` key and if there is just resolve its ObjRef.

Does this make sense? I fear this approach would need some major work...

Maybe I'll try a custom class in the weekend to see how much work it is.

typhoon71 replied:
Method extracted.

pietermarsman replied:
I just ran the test suite and noticed that it took 40 seconds to parse _nonfree/nlp2004slides.pdf_. Maybe we can use that to see if we can improve the speed. Different issue / PR though ;)

pietermarsman commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+        parser = PDFParser(fp)
+        
+        doc = PDFDocument(parser)
+        res = resolve1(doc.catalog)
```
Maybe add a paragraph about what the `doc.catalog` is (because I don't know :p ).

typhoon71 replied:
How do I add stuff to or modify a PR? 
(I can add a paragraph that explains doc.catalog later today)

jstockwin replied:
You can simply add new code, add a new commit, and push to your fork's `develop` branch (since that's what you started this PR from), and this PR will update.

This PR will always be the difference between your `develop` branch and the pdfminer one.

typhoon71 replied:
Nice, thanks for the info.

jstockwin commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+
+The second thing you need is a PDF with AcroForms (as found in PDF files with fillable forms or multiple choices).
+
+It won't work for XFA forms.
```
Just for completeness, do you know if pdfminer.six can handle XFA forms (but in a different manner)?

I appreciate you might not have experience or interest in these forms, but we don't want to make it sound like pdfminer can't handle something if it can.

typhoon71 replied:
I don't think XFA forms can be handled by pdfminer, and sadly I got no experience with them.

There's an article on SO: https://stackoverflow.com/questions/55754342/.

I understand that it's non trivial to add XFA support.

jstockwin commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+        doc = PDFDocument(parser)
+        res = resolve1(doc.catalog)
+
+        if 'AcroForm' in res:
```
Maybe this is cleaner as e.g.
```
if 'AcroForm' not in res:
    raise ValueError("No AcroForm Found")
```
since then it's an obvious message about the problem, and we can drop an indentation level below.

What do you think?

typhoon71 replied:
Will do, makes sense to me. 
(my original code had a couple of other conditions,not relevant here, so I just removed them and the it was left like that)

typhoon71 replied:
Done.

jstockwin commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+
+                if value:
+
+                    # resolve indirect obj
+                    while isinstance(value, PDFObjRef):
+                        value = resolve1(value)
+
+                    # decode PSLiteral, PSKeyword
+                    if isinstance(value, (PSLiteral, PSKeyword)):
+                        value = value.name
+
+                    # decode bytes
+                    if isinstance(value, bytes):
+                        value = utils.decode_text(value)
```
Not for this PR, but in general it seems silly to me that users have to hope with different types here to decode. Perhaps we should add a util function that does all these if statements for you? (cc @pietermarsman do you think that's worth an issue?)

pietermarsman replied:
It could make sense. There are already some utility functions for this in _pdftypes.py_: `resolve1()`, `resolve_all()`, `decipher_all()` and methods for resolving specific types. 

Not sure if `resolve1` -> `get name` -> `decoding` is a common pattern though.

typhoon71 replied:
I don't know if this is common, but my form needed it because I could get those types as a value... or combinations of them, like value as an indirect object which resolved into a byte encoded PSLiteral... no idea how one does that.

The key were better, they just needed proper decoding (being byte). 

Btw: what does decipher_all() do?

jstockwin commented on docs/source/howto/index.rst:
```diff
@@ -9,3 +9,4 @@ How-to guides help you to solve specific problems with pdfminer.six.
     :maxdepth: 1
 
     images
+    aforms
```
Just to keep things nice and explicit, I'd rename the file to `acro_forms`.

typhoon71 replied:
Will do

typhoon71 commented:
> Finally, one question: What happens if there are two separate forms on the PDF? Or is that not something that can happen?

As far I know there shoud be onluy one Acroform dict even if there are more forms on different pages (everything will go in the same data structure). Th PDF files I work with range from 200/300 to 700 form fields.

Btw, what do you think it's better?

```
while isinstance(value, PDFObjRef):
    value = resolve1(value)
```
or
```
if isinstance(value, PDFObjRef):
    value = resolveall(value)
```
The second method is available but I'm not sure which is better.

typhoon71 commented:
Updated the howto with some extra info.

Added a get_acroform() method to pdfdocument class, it returns all field data in a dict
Added a PDFAcroFormError for completion too.
Spotted a bug in my code, that would be triggered when a field value was b'' (empty byte), so I fixed it.

If the method is fine, would it make sense to document that and remove everything else?
This became less documentation that I supposed it to. XD

I will see what I can do for AcroForms that contain a tree of forms, like in https://www.federalreserve.gov/reportforms/forms/FR_Y-1520170331_f.pdf, but later this week. 
I never found one of those but I suppose I can work on it too.

pietermarsman commented on pdfminer/pdfdocument.py:
```diff
@@ -812,3 +816,24 @@ def read_xref_from(self, parser, start, xrefs):
         return
+    
+    # get AcroForm data
+    def get_acroform(self):
```
I meant to extract the method and use it in the docs. People that copy-paste a method are better of than people than people that copy-paste a script. (At least, that's my intuition). 

The method looks good though :) Perhaps we can add it to the `highlevel` methods and rename it to `extract_form_data()`.

typhoon71 replied:
Oh, I misunderstood.
I'll remove the method from pdfdocument, and change the howto.
It won't take muck for me to make it again for the highlevel later on.

typhoon71 replied:
I see the pdfdocument in my fork is different (there was a release): what do I need to do?

pietermarsman commented on docs/source/howto/aforms.rst:
```diff
@@ -0,0 +1,56 @@
+
+                    # resolve indirect obj
+                    while isinstance(value, PDFObjRef):
+                        value = resolve1(value)
```
I think this can be replaced by just `value = resolve1(value)`. The `resolve1` method already has a while loop in it.

typhoon71 replied:
You're right, no sense repeating both the while loop and the isinstance check.

pietermarsman commented:
> Btw, what do you think it's better?

I think `value = resolve1(value)` (without while) does the job. You only need `resolve_all` if you expect the output to be a list or dict, which could be the case with the hierarchical AcroForm's.

Let's focus this PR on the non-hierarchical forms. We can always add it later on.

pietermarsman commented:
> If the method is fine, would it make sense to document that and remove everything else?

I meant using the method in the docs, not in the code. Not sure if it makes sense in the code base. It probably needs a few iterations if we want to add it there to cover all the edge cases. For the docs its fine if it can handle the example PDF (and a fair amount of others) well.

jstockwin commented on docs/source/howto/acro_forms.rst:
```diff
@@ -0,0 +1,124 @@
+            # decode name
+            name = decode_text(name)
+
+
```
```suggestion
```

jstockwin commented on docs/source/howto/acro_forms.rst:
```diff
@@ -0,0 +1,124 @@
+
+    res = resolve1(doc.catalog)
+
+- Check if the catalog contains the AcroForm key and raise ValuError if not 
```
```suggestion
- Check if the catalog contains the AcroForm key and raise ValueError if not 
```

jstockwin commented on docs/source/howto/acro_forms.rst:
```diff
@@ -0,0 +1,124 @@
+    res = resolve1(doc.catalog)
+
+- Check if the catalog contains the AcroForm key and raise ValuError if not 
+(the PDF does not contain Acroform type of ibteractive forms if this key is missing in the catalog, see section 12.7.2 of PDF 32000-1:2008 specs)
```
```suggestion
(the PDF does not contain Acroform type of interactive forms if this key is missing in the catalog, see section 12.7.2 of PDF 32000-1:2008 specs)
```

typhoon71 commented:
If I understand correctly, to complete this I need:

- [ ] Create a method out of the howto. I'm not too sure that it's needed, also I'd like to ask what the method should cover, ie just field extraction while file opening would be another method? (2 small methods).
- [ ] Provide test PDF files (I can't share company stuff, sorry). The code works for all my PDFs (20k+ forms of variuos type) but it's just my word... 
- [ ] CHANGELOG.md description? Is that needed?

pietermarsman commented:
> Create a method out of the howto.

Only if you feel like it. 

> Provide test PDF files

This I would really like. Makes the howto easier to reproduce. 

> CHANGELOG.md description? Is that needed?

Nope.

pietermarsman commented:
Maybe you can use this file (or only the first page).

Originally from https://stackoverflow.com/questions/58071962/pdf-file-to-dict-returns-weird-characters/58090542#58090542.

[Original_CS.pdf](https://github.com/pdfminer/pdfminer.six/files/4968113/Original_CS.pdf)

typhoon71 commented:
- Regarding making a method of the howto, I think I won't; I think it should stay simple like it is: also, it will be easier to compare it to the current examples on SO and, not lastly, to the recursive field resolver (WIP). Making a method out of it is trivial anyway.

- You just linked to the file I found yesterday reading #293, and tested on the fly. Checkthe issue, it's solved... well, there was no issue. My question is: is it fine to include files here taking it from the net?

typhoon71 commented:
I added a couple of PDF files with AcroForm to the sample: it's just a quick example with the basic items, but it shows that the howto code is functional, both on a clean and on a compiled form.

I took the occasion to implement a case of value resolving that I did not encounter before, the ComboBox.
In this case the previous code would return a list but the values inside it not be decoded, now it's fixed.

Regarding the files in the article (https://stackoverflow.com/questions/58071962/pdf-file-to-dict-returns-weird-characters/58090542#58090542), I don't feel confortable taking files I don't own an putting them here, so I won't be adding it to the samples myself; feel free to add it tough, both the original one and the edited one are good examples/test for the howto.

If everything is fine I say this could be merged?

jstockwin commented on docs/source/howto/acro_forms.rst:
```diff
@@ -0,0 +1,145 @@
+
+Before you start, make sure you have :ref:`installed pdfminer.six<install>`.
+
+The second thing you need is a PDF with AcroForms (as found in PDF files with fillable forms or multiple choices).
```
Now we've included some PDFs, let's mention them here? Something like:

```suggestion
The second thing you need is a PDF with AcroForms (as found in PDF files with fillable forms or multiple choices). There are some examples of these in the GitHub repository under `samples/acroform`.
```

typhoon71 commented:
@jstockwin 
Regarding pdfdocument.py diffs: they were caused by the recent PDFTextExtractionNotAllowedError issue, they weren't merged in my fork. Now it's fixed.

Btw: how is one supposed to merge changes made to the main repo into a fork one?

jstockwin commented:
@typhoon71 looks good, although I think there is still a diff with the pdfdocument file?

You can run `git fetch upstream` to update your local copy of `pdfminer/pdfminer.six` branches (specifically the `develop` branch). You should then either rebase (preferred but perhaps more complicated if you don't know what you're doing) with `git rebase upstream/develop` or merge with `git merge upstream/develop`. That get you the new changes into your branch.... you may end up with conflicts merge conflicts that you have to resolve.

Before doing anything, check out some backup branch so if you mess it up you can always get back to it.... (`git checkout -b develop-backup-branch` to create the new branch and then `git checkout develop` to go back to your branch).

Git can be quite complicated and I don't want you to mess up all your hard work following my commands, so please make sure you understand them. Alternatively, @pietermarsman might be able to solve the conflicts for you... not sure.

typhoon71 commented:
> 
> 
> @typhoon71 looks good, although I think there is still a diff with the pdfdocument file?
> 
> You can run `git fetch upstream` to update your local copy of `pdfminer/pdfminer.six` branches (specifically the `develop` branch). You should then either rebase (preferred but perhaps more complicated if you don't know what you're doing) with `git rebase upstream/develop` or merge with `git merge upstream/develop`. That get you the new changes into your branch.... you may end up with conflicts merge conflicts that you have to resolve.
> 
> Before doing anything, check out some backup branch so if you mess it up you can always get back to it.... (`git checkout -b develop-backup-branch` to create the new branch and then `git checkout develop` to go back to your branch).
> 
> Git can be quite complicated and I don't want you to mess up all your hard work following my commands, so please make sure you understand them. Alternatively, @pietermarsman might be able to solve the conflicts for you... not sure.

The files should be identical, I pasted the whole pdfdocument from pdfminer in my fork; maybe it just shows the "update"?

jstockwin commented:
Hm, I'm not sure... https://github.com/pdfminer/pdfminer.six/pull/458/files#diff-7b8996b6e630ef8e86e17a8d0df3c78c shows a diff to me so I think something is wrong. Let's wait for @pietermarsman as he might be able to fix that for you anyway... I'm not 100% sure how permissions etc work with forks.

typhoon71 commented:
> Hm, I'm not sure... https://github.com/pdfminer/pdfminer.six/pull/458/files#diff-7b8996b6e630ef8e86e17a8d0df3c78c shows a diff to me so I think something is wrong.

Sorry, I copied the wrong one (from another branch). Now it's the good one.
I'm not familiar with github publishing, and it shows... I didn't even notice the " Files changed" tab!

pietermarsman commented:
@jstockwin @typhoon71 thanks for all the work! Sorry it took me a while to merge it.
